signac/core/collection_api.py

```diff
@@ -0,0 +1,446 @@
+from collections.abc import MutableSequence
+
+try:
+    import numpy
+    NUMPY = True
+except ImportError:
+    NUMPY = False
+
+try:
+    from collections.abc import Collection
+except ImportError:
+    # create Collection
+
+    from collections.abc import Sized, Iterable, Container
+
+    def _check_methods(C, *methods):
+        mro = C.__mro__
+        for method in methods:
+            for B in mro:
+                if method in B.__dict__:
+                    if B.__dict__[method] is None:
+                        return NotImplemented
+                    break
+            else:
+                return NotImplemented
+        return True
+
+    class Collection(Sized, Iterable, Container):  # type: ignore
+        @classmethod
+        def __subclasshook__(cls, C):
+            if cls is Collection:
+                return _check_methods(C,  "__len__", "__iter__", "__contains__")
+            return NotImplemented
+
+        @classmethod
+        def __instancecheck__(cls, instance):
+            for parent in cls.__mro__:
+                if not isinstance(instance, parent):
+                    return False
+            return True
+
+
+class SyncedCollection(Collection):
+
+    def __init__(self, parent=None):
+        self._data = None
+        self._suspend_sync_ = 0
+        self._parent = parent
+
+    @classmethod
+    def __instancecheck__(cls, instance):
+        if not isinstance(instance, Collection):
+            return False
+        else:
+            return all(
+                [hasattr(instance, attr) for attr in
+                 ['sync', 'load', 'to_base', 'from_base']])
+
+    @classmethod
+    def from_base(self, data, parent=None):
+        if isinstance(data, Mapping):
+            return SyncedDict(data, parent)
+        elif isinstance(data, Sequence) and not isinstance(data, str):
+            return SyncedList(data, parent)
+        elif NUMPY:
+            if isinstance(data, numpy.number):
+                return data.item()
+            elif isinstance(data, numpy.ndarray):
+                return SyncedList(data.tolist(), parent)
+        return data
+
+    def to_base(self):
+        pass
+
+    @contextmanager
+    def _suspend_sync(self):
+        self._suspend_sync_ += 1
+        yield
+        self._suspend_sync_ -= 1
+
+    def _load(self):
+        pass
+
+    def _sync(self):
+        pass
+
+    def sync(self):
+        if self._suspend_sync_ <= 0:
+            if self._parent is None:
+                self._sync()
+            else:
+                self._parent.sync()
+
+    def load(self):
+        if self._suspend_sync_ <= 0:
+            if self._parent is None:
+                self._load()
+            else:
+                self._parent.load()
+
+
+class _SyncedDict(SyncedCollection, MutableMapping):
+
+    VALID_KEY_TYPES = (str, int, bool, type(None))
+
+    def __init__(self, data=None, parent=None):
+        super(_SyncedDict, self).__init__(parent=parent)
```

Comment:
Python2 is officially gone, so I would use plain `super()` wherever possible. However, in this package you will eventually have to start being careful about that since we're creating diamonds all over the place. Once `SyncedDict` and `JSONCollection` become abstract, for instance, you'll want to make sure that if they require initialization that `JSONDict` contains the appropriate calls to `super(PARENT_TYPE, self).__init__(...)`.